### ISSUE TITLE
Virtual indentation of multiline text strings. (Editor Feature)

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#73661]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Virtual-indentation-of-multiline-text-st/10659683)._

---
Multi-line text in C# and VB.NET do not align with existing code, in many cases significantly reducing readability.
Please modify the Visual Studio code editor to virtually indent this text to improve alignment.  This indentation should not be saved in the code file, and it would be nice if the IDE grayed out the virtual indentation area to indicate this.
In this example, the indentation is included within the text string to improve code readability (imagine a much larger SELECT statement with multiple joins and criteria, for the first, and a larger message for the second), but it truly just adding extra, useless data to the string.  If that indentation was virtual instead, it would maintain the readability without the useless spaces being included.  For the SELECT statement, the addition of the excess spaces does not adversly affect the SELECT command, but it would cause more data to be sent to the SQL Server compared to what is needed.  For the Msg, however, the excess spaces cause the text to be displayed incorrectly where it is used.
```
            Dim SQL As String
            Dim Msg As String

            SQL = "SELECT WidgetID, VendorID, Name, Description
                   FROM Widgets
                   WHERE (Name LIKE '%custom%') AND (VendorID = 52)" 'VendorID 52 = Jane's house of widgets.

            Msg = "This is a list of widgets with their description.
                   These are widgets that can be acquired from Jane's house of widgets.
                   Note that only custom widgets are shown here."

            ...
```

Note that a previous suggestion was closed as a duplicate of [Design a formatter extension API](https://github.com/dotnet/roslyn/issues/31691), however that seems to be a generic issue about redesigning the formatting system, and appears to be going nowhere, as such it does not seem to be equivalent.

---
#### Feedback Bot on 5/16/2024, 02:40 AM: 

(private comment, text removed)

**Comment by mdell-seradex** [comment]
I haven't seen this in C#, and find it quite disappointing that it won't be supported in VB.
Does it have to be enabled somewhere, or does it only work with certain project types?

This is really something that can be language agnostic.

**Comment by mdell-seradex** [comment]
> This is already supported in C# for raw-string-literals (this exact use case is part of hte language). We are not doing anything in this space for VB (either at the language or tooling level).

@CyrusNajmabadi, I haven't seen this in C#, and find it quite disappointing that it won't be supported in VB.
Does it have to be enabled somewhere, or does it only work with certain project types?
Is there a setting I need to enable to get it to work?

This is really something that can be language agnostic.

**Comment by CyrusNajmabadi** [comment]
> Does it have to be enabled somewhere

You have to be on c#11 to use them.

**Comment by CyrusNajmabadi** [comment]
> This is really something that can be language agnostic.

The position of the .net team is that the vb language will not be getting any more changes.  We'll update its compiler to handle the latest .net constructs.  But we're not changing the language specification for it.

**Comment by CyrusNajmabadi** [comment]
> It would be great if this was available everywhere.

Understood.  We've made the decision to not do this for vb.  Sorry. 

> but you say that this new behaviour is only in C# 11.

Correct. Multi line RAW string literals are a new c#11 feature.  

> It will definitely take a significant amount of time, likely years, before we can move our entire codebase to .NET 6 and above.

You can upgrade your c# version without upgrading your .net version.  Raw string literals will work with any target. They don't need a .net upgrade.

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, my understanding is that we can't use C# 11 with .NET Framework 4.8, yet, multi line RAW string literals work in .NET Framework projects.  We have several examples of them now in our code and they compile correctly.
Perhaps if the new SDK project type (I think that is what it is called) was used, but I am not aware of any existing method via the tools to convert an existing project to that type, and it seems that it is not planned for those project types to completely support .NET Framework in Visual Studio.

**Comment by CyrusNajmabadi** [comment]
> yet, multi line RAW string literals work in .NET Framework projects.

Yes.  That is correct.  raw string literals have no special runtime requirements.  They're just normal System.Strinsg.

> We have several examples of them now in our code and they compile correctly in both C# and VB.

VB does not have raw-string literals.

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, it may not technically have "RAW" string literals, but we can do multiline strings in VB.
This is just one example.  Note that while this uses string interpolation, we have other examples that do not.
![image](https://github.com/dotnet/roslyn/assets/40300659/815cf707-0e86-49ca-aed3-3696e9babe8e)

**Comment by mdell-seradex** [comment]
> The position of the .net team is that the vb language will not be getting any more changes. We'll update its compiler to handle the latest .net constructs. But we're not changing the language specification for it.
Isn't this Visual Studio editor formatting, and not language syntax/functionality?  I wouldn't have thought that is related to the "language specification", but I guess I can see how they can be intertwined.  I suppose it can be difficult to properly determine on what side an issue lies.

**Comment by CyrusNajmabadi** [comment]
> unfortunately, the solution you have suggested is NOT available for us to use

it should be.  This feature doesn't utilize anything not available on .net framework projects.  As an example, we use it on roslyn, which still targets .net framework (since it runs in VS, which is still a .net framework app).  We use raw strings everywhere.

> Our desktop application involves upwards of 100 projects that run under a single process. Converting to a multi-process application 

You don't need to be a multi-process application to use raw strings.

> Our desktop application involves upwards of 100 projects that run under a single process. 

THat's the same as Roslyn.  We're roughly 250 projects (and around 1000 virtual projects), almost all of which run in a single process.

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, you said RAW strings were introduced in C# 11.  From everything I know, we cannot use C# 11 in .NET Framework 4.8 projects.  I said multi-process because, as I understand it, we would need to target .NET 6 or higher to use C# 11, and using .NET 6 and .NET Framework 4.8 in a single process is not supported.  From what I know, C# 8 is the last version for .NET Framework.
I mentioned the number of projects to give you a sense of what it would take for us to convert them all to .NET 6+.
We are not a large company like Microsoft, where we could put 100 employees on converting these without affecting day-to-day operations (we don't have 100 employees).

**Comment by CyrusNajmabadi** [comment]
> From everything I know, we cannot use C# 11 in .NET Framework 4.8 projects.

Raw strings work fine in 4.8 projects.  As mentioned, we're already using them in roslyn and we target exactly that.  Raw strings have no runtime dependency associated with them whatsoever. :)

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, are you saying that you are not using SDK-style projects in roslyn for 4.8, or did you manually retarget them to 4.8, or did I not see some official Microsoft documentation about making that work.
Most of our projects are NOT the newer SDK-style, and the only thing I have seen about converting them will upgrade them to the newer .NET version.

One big unfortunate thing for us is that we have many VB projects.  It was extremely disheartening when I heard many years ago that Microsoft wanted to essentially abandon an excellent language, than is actually better than C# in many respects.

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi or are you saying that when RAW strings were introduced in C# 11, they were backported to previous C# versions in the supported IDEs?  I feel a bit lost.

**Comment by CyrusNajmabadi** [comment]
You need a modern VS to get support for raw-string-literals.  But the feature works downlevel just fine.  The only thing required is that you set your lang version to C# 11 or above.   There is no runtime requirement or restriction about using raw strings with .net framework.

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, most of us are using Visual Studio 2022, and when I look to change the C# version, I see that it is "Automatically selected based on framework version".  It gives me a link [Why can't I select a different C# version?](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#edit-the-project-file), but that directs me to a page that doesn't tell me why.  Reviewing the version of that page on github from prior to May 7th, 2024, it says that changing the version of C# is NOT recommended because it will allow using features not supported by the version of .NET that the project targets, leading to hard to diagnose issues.  The current version of that page, and the pages it points to are not as clear about this.

**Comment by CyrusNajmabadi** [comment]
In your case only use features supported in framework, not core. The compiler will error if you try to use an unsupported feature

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, exactly, which says to me that we can't use RAW strings, as it was added in C# 11, but .NET Framework uses C# 7.3 (not 8 like I previously thought), thus I am stuck back at not having virtual indentation.

**Comment by CyrusNajmabadi** [comment]
I was saying the opposite.  You can use raw string literals.  Not sure how much more explicit i can be on this point :)

> Microsoft Copilot told me that as well.

It is incorrect (possibly hallucinating).  Don't trust AI for accurate answers on topics.  Always verify it.

**Comment by CyrusNajmabadi** [comment]
![image](https://github.com/dotnet/roslyn/assets/4564579/fa35127b-fe7a-4597-a0cc-19445da05d86)

**Comment by mdell-seradex** [comment]
That is exactly the kind of behaviour I'd like. I can see the adjusted margin too which is nice. It's just too bad that it's only for raw string literals.
Note that even the documentation says it is for C#11, unless I am misunderstanding it. [Raw string literal text](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/raw-string). I will try it anyway.
I wonder if I could create an extension to add this behaviour for verbatim strings...

**Comment by CyrusNajmabadi** [comment]
> It's just too bad that it's only for raw string literals.

It's only for raw-string-literals because indentation is part of the design of raw-srting-literals.  If you want this, switch your strings to raw strings.  This is *explicitly* why we created these:)

> Note that even the documentation says it is for C#11

That is correct.

> I wonder if I could create an extension to add this behaviour for verbatim strings...

if you want.  but raw-string-literals are strictly superior to multi-line verbatim strings.

**Comment by mdell-seradex** [comment]
I understand that and will definitely recommend it where we can use it, but our existing code, until it is converted, and everything in VB can benefit from it.

**Comment by CyrusNajmabadi** [comment]
Good luck :)

**Comment by mdell-seradex** [comment]
@CyrusNajmabadi, I was finally able to test this, and I don't know how you built your .NET Framework project, but unless you forced it to use C# 11 which is not recommended, it does not work.
![image](https://github.com/dotnet/roslyn/assets/40300659/bb25871b-9b04-42fe-8960-ec841901486a)

Were you saying to forcibly change the project to use C# 11, but to manually try not use features that aren't supported?  That seems like a risky idea.
Based on the documentation, it sounds like the IDE will suggest things that aren't supported, and neither it, nor I, would know until attempting to compile and run the assembly.

**Comment by CyrusNajmabadi** [comment]
Yes. In order to use features from a particular version of the language, I told the compiler to use that version that I wanted to use. 

> Based on the documentation, it sounds like the IDE will suggest things that aren't supported

What documentation are you referring to. 

The IDE will only suggest things supported by your language version and particular runtime you are targeting.

> , and neither it, nor I, would know until attempting to compile and run the assembly

That's not how Roslyn works. We'll error live if you try to use something not supported (like dims). We'll also error when you compile. You don't have to wait till runtime to find out. 

The only feature that waits till runtime to fail is `dynamic`.  Because, well... that's the point of it :-)

**Comment by mdell-seradex** [comment]
This page [C# Compiler Options for language feature rules](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/language#langversion) says this:

> Important
> 
> The latest value is generally not recommended. With latest, the compiler enables the latest features, even if those features depend on updates not included in the configured target framework.

To me that implies that the same thing would happen when specifying a version that is higher than what is supported by the version of .NET that I am using.

Also, the version of this page [Configure C# language version](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#edit-the-project-file) from [April 16th, 2024](https://github.com/dotnet/docs/blob/df7c9346491dec7dbdd9ab9449db1d80e7749e0d/docs/csharp/language-reference/configure-language-version.md), says this:
- > The choice of default ensures that you use the latest language version compatible with your target framework.
- > Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors.
- > [C# 11](https://github.com/dotnet/docs/blob/df7c9346491dec7dbdd9ab9449db1d80e7749e0d/docs/csharp/whats-new/csharp-11.md) is supported only on .NET 7 and newer versions

Note that this documentation was changed to remove these as part of [Fix C# reference TOC for offline book #40794](https://github.com/dotnet/docs/pull/40794), however I can see no explanation, or even a hint, as to why these apparently important notes were removed.  As such I take that change as wrong.

**Comment by CyrusNajmabadi** [comment]
> As such I take that change as wrong.

You do you.  Fundamentally, you make the choices here about yoru codebase.  I'm simply explaining how these features work and what your options are.  If you choose to not use them, c'est la vie.

**Comment by mdell-seradex** [comment]
> @mdell-seradex that is referring to `<LangVersion>latest</LangVersion>`. Which then automatically changes your lang version as you update your compiler to newer compiler versions. That's not what i was recommending.

I understand that, but it does imply that the same thing can happen when I target a version of C# that my version of .NET doesn't support.

**Comment by CyrusNajmabadi** [comment]
I have no idea what that is referring to. As I stated, we will tell you live, and at compile time, if you use something not supported (unless you are using dynamic). 

I don't know what to tell you.  You make the decision about your code base.  I've explained how the language, compiler, and features here work. Ultimately it's up to you if you use them or if your concerns outweigh the explanations. 

Fundamentally, we used all the info when designing these language features and delivering support for them. And, add I've Said a few times, we use all these language features ourselves (thousands and thousands of times over), despite being net framework itself). 

I genuinely don't know what else I can say about this. Reading back it seems like this has been stated several dozen times with no new info coming up. 

So unless there is something new you want to discuss, I'll simply wish you good luck on this. If there is something new, I'd also recommend visiting the discord so you can just talk to the team directly.

**Comment by mdell-seradex** [comment]
Unless I have something else to say, I won't comment further, or post to discord, although I don't know how to visit the discord to speak to the team.

**Comment by CyrusNajmabadi** [comment]
You're welcome to comment further.  I'm just saying that I don't know how many more ways to assure you on this particular topic. 

The discord information is on our main page. Thanks!

**Comment by mdell-seradex** [comment]
I see it now (https://github.com/dotnet/roslyn/).  Hmm, the image links seem to be broken.
